### PR TITLE
Fix for #1155 thenThrow(Class) 

### PR DESCRIPTION
**Pull request by ChristianSchwarz**
Fix for #1155 
Mockito.thenThrow(Class) now fails when the given Throwable type is not declared on the stubbed method.

### CODE REVIEW

**Comment by ChristianSchwarz** [comment]
The build fails for JDK7 seem like an CI problem: Execution failed for task ':fetchAllContributors'.

**Comment by szczepiq** [comment]
>The build fails for JDK7 seem like an CI problem: Execution failed for task ':fetchAllContributors'.

I will look into that, thank you.

**Comment by ChristianSchwarz** [comment]
@TimvdLippe could you please remove the checkstyle rule [NewlineAtEndOfFile] it is frustration to fix and adds zero value

**Comment by TimvdLippe** [comment]
If you have an editor with editorconfig support, it is no manual work. Both Eclipse and Intellij support it. Intellij even right out of the box. Else you can manually add the newline. I think you configured your editor to revert that change somewhere in your preferences.

Anyways, editorconfig makes sure we avoid unnecessary noise when contributing, even if you are new to the project and unaware that it exists. Its thus more friendly to new developers and I dont want to regress in that regard :(

**Comment by ChristianSchwarz** [comment]
The IDE added a newline, also I tried to add manually a newline, but check style still fails. I have no further time to investigated into it.
----
This PR fixes now cleans up the unfinished stubbing state  in case null was passed to thenThrow(Class)/doThrow(Class). This leaded to unpredictable results in test suites before.

**Comment by TimvdLippe** [comment]
You are fixing the wrong files. It is StubbingWithThrowables test: https://travis-ci.org/mockito/mockito/jobs/266766666#L303

**Comment by ChristianSchwarz** [comment]
Strange when I run it locally it complains about the 3 files of the last commit...

File: src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java [hunk 1]
```diff
@@ -90,12 +95,9 @@ public void shouldAllowSettingCheckedException() throws Exception {
 
         when(reader.read()).thenThrow(ioException);
 
-        try {
-            reader.read();
-            fail();
-        } catch (Exception e) {
-            assertEquals(ioException, e);
-        }
+        exception.expect(sameInstance(ioException));
```
Comment (TimvdLippe) [inline src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Missing the ewuals here

File: src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java [hunk 2]
```diff
@@ -39,48 +52,40 @@ public void shouldStubWithThrowable() throws Exception {
...
 
         doThrow(expected).when(mock).clear();
-        try {
-            mock.clear();
-            fail();
-        } catch (Exception e) {
-            assertEquals(expected, e);
-        }
+
+        exception.expect(IllegalArgumentException.class);
```
Comment (TimvdLippe) [inline src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Missing equals here

File: src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java [hunk 3]
```diff
@@ -39,48 +52,40 @@ public void shouldStubWithThrowable() throws Exception {
         IllegalArgumentException expected = new IllegalArgumentException("thrown by mock");
         when(mock.add("throw")).thenThrow(expected);
 
-        try {
-            mock.add("throw");
-            fail();
-        } catch (IllegalArgumentException e) {
-            assertEquals(expected, e);
-        }
+        exception.expect(sameInstance(expected));
```
Comment (TimvdLippe) [inline src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Missing equals here

Comment (ChristianSchwarz) [reply src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Can you give me some more infos? We expect the instance to be thrown that was passed to the stubbing, right?

File: src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java (hunk 2 above)
Comment (ChristianSchwarz) [reply src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Lets better use: `exception.expect(sameInstance(expected));`

File: src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java (hunk 3 above)
Comment (TimvdLippe) [reply src/test/java/org/mockitousage/stubbing/StubbingWithThrowablesTest.java]: Sorry, this one is correct. I misread

**Comment by TimvdLippe** [comment]
@ChristianSchwarz Nice, thanks! While could seems good, could you add a test (could just copy the literal test from #1155) to also check that case? I couldn't spot it in the added tests.

**Comment by ChristianSchwarz** [comment]
Sure I can add more tests, but what you mean by "that case"?

**Comment by TimvdLippe** [comment]
`when(mock.doIt()).thenThrow(Exception.class);` E.g. with only a literal class passed on, expecting this to fail.

**Comment by ChristianSchwarz** [comment]
I am not sure what you mean by "only a literal class passed on, expecting this to fail".

Maybe a stubbing with an checked Exception type where an other checked exception is declared, like this:

```
class ExceptionA extends Exception

class ExceptionB extends Exception

public void doIt() throws ExceptionA

when(mock.doIt()).thenThrow(ExceptionB.class) 
//^^ fails like "can't throw checked exception >ExceptionB< cause it is not declared"
```

**Comment by TimvdLippe** [comment]
The original report states that the following test did not fail. I would like to see that literal test as a regression test in our test body:

```java
@Test(expected = MockitoException.class)
    public void using_thenThrow_Class() {
        C mock = mock(C.class);
        // Set a checked exception that does not matche the method signature
        when(mock.doIt()).thenThrow(Exception.class);
    }
```

**Comment by TimvdLippe** [comment]
@ChristianSchwarz Thanks for all your work!

**Pull request merged**